### ISSUE TITLE
Bad performance in comparison to native https.get

### ISSUE THREAD

**Issue by max-mykhailenko** [bitinn/node-fetch#405]
Run `node test.js https://google.com` and get
```
Execution time NATIVE: 0s 150.41529ms
Execution time FETCH: 0s 341.750908ms
```
Overhead about 50% (O_O)

Save code as `test.js`
```js
require('isomorphic-fetch');
let start;

start = process.hrtime();
fetch(process.argv[2], {
  method: 'GET',
}).then(() => {
  const hrend = process.hrtime(start);
  console.log(`Execution time FETCH: ${hrend[0]}s ${hrend[1] / 1000000}ms`);
});


const https = require('https');
start = process.hrtime();
https.get(process.argv[2], resp => {
  let data = '';

  resp.on('data', chunk => {
    data += chunk;
  });

  resp.on('end', () => {
    const hrend = process.hrtime(start);
    console.log(`Execution time NATIVE: ${hrend[0]}s ${hrend[1] / 1000000}ms`);
  });
});
```

How I can improve performance but keep using node-fetch?

**Comment by bitinn** [comment]
first of all, using google.com as your target is a terrible idea, if you do want to test raw script performance, isolate network performance, run a local server and test that (we got server.js in the /tests/ folder if you need one, or use something like [serve](https://github.com/zeit/serve) to setup a local server).

secondly, I suspect any slowdown you will get is inherit to native promise startup time (you basically spin up a node worker thread for your test script, unlike in real application, you are constantly running). `node-fetch` does so little extra processing, a 150ms slowdown is highly unlikely, unless you got a ancient cpu.

thirdly, assuming in a production environment, any slowdown from network lib will probably be insignificant to the actual network latency. if you want raw performance then unlikely fetch will match native http but I suspect they won't be too far off.

TL;DR: you need a better test, we would like to see one too, but `node-fetch` is probably good enough for most cases.

**Comment by jimmywarting** [comment]
Made me curious of how fast [request](https://www.npmjs.com/package/request) is also

**Comment by max-mykhailenko** [comment]
I tried `request` too and results very similar to `node-fetch` when I ping google (from cloud servers). BUT requests to my API shows almost same time for all libs. What do you think?

**Comment by bitinn** [comment]
@max-mykhailenko I think it's not a network library problem.

**Comment by bitinn** [comment]
And by the way, in case you don't know connection pooling #196

**Comment by max-mykhailenko** [comment]
I understand where was a problem in my case: balancer, php API and several node process was on one server and when I tried to test performance node and php didn't leave cpu resources for balancer and every next request takes more and more time for balancing.

Thanks for participating!
